File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/persistence/HttpSoulManifestStore.kt
package org.wyrdsekai.app.engine.persistence

import org.wyrdsekai.app.engine.soul.ClientSoulManifest
import org.wyrdsekai.app.network.SoulClient

/**
 * [SoulManifestStore] backed by the server's soul REST API via [SoulClient].
 *
 * This enables client-side soul sync (Gap 5): the phone's soul manifest is
 * persisted to the household server so that other nodes (desktop, other phones,
 * server-side agents) can see the latest state.
 *
 * Operations:
 * - [save] syncs the manifest to the server via POST
 * - [load] fetches the latest manifest from the server via GET
 * - [delete] is a no-op (the server does not expose a delete endpoint yet)
 * - [listDids] returns an empty list (would need a server listing endpoint)
 *
 * For offline-first resilience, pair this with a local store (e.g.
 * [InMemorySoulManifestStore] or platform-specific file store) and use
 * this HTTP store as a sync target, not the sole source of truth.
 */
class HttpSoulManifestStore(
    private val soulClient: SoulClient,
    private val token: String,
) : SoulManifestStore {

    /**
     * Sync the manifest to the server. Throws on network failure so
     * callers (typically CompanionEngine) can catch and retry later.
     */
    override suspend fun save(manifest: ClientSoulManifest) {
        soulClient.syncManifest(manifest.did, manifest, token).getOrThrow()
    }

    /**
     * Fetch the latest manifest for the given DID from the server.
     * Returns null if the server has no manifest for this DID or on error.
     */
    override suspend fun load(did: String): ClientSoulManifest? {
        return soulClient.getLatest(did, token).getOrNull()
    }

    /**
     * No-op — the server does not expose a delete endpoint yet.
     * When it does, this will call DELETE /api/soul/{did}.
     */
    override suspend fun delete(did: String) {
        // Server delete endpoint not yet available
    }

    /**
     * Returns an empty list — the server does not expose a listing endpoint.
     * When it does, this will call GET /api/soul and return all DIDs.
     */
    override suspend fun listDids(): List<String> = emptyList()
}

File: clients/rn/src/engine/persistence/HttpSoulManifestStore.ts
import type { ClientSoulManifest } from '../soul/SoulManifest';
import type { SoulManifestStore } from './SoulManifestStore';
import { getLatestManifest, syncManifest } from '../../network/SoulClient';

/**
 * #7 (2026-07-19 OSS hardening) — a {@link SoulManifestStore} backed by the
 * server's soul REST API. Used as a SYNC TARGET (not the sole source of truth):
 * the phone keeps a local store and additionally pushes evolved manifests here
 * so phone-side soul evolution (PhoneForge on sleep) reaches the household.
 *
 * TS port of the KMP HttpSoulManifestStore. save() is best-effort — the
 * underlying syncManifest returns null on failure rather than throwing, so a
 * network error is silently retried on the next sleep.
 */
export class HttpSoulManifestStore implements SoulManifestStore {
  constructor(
    private readonly serverUrl: string,
    private readonly token?: string,
  ) {}

  async save(manifest: ClientSoulManifest): Promise<void> {
    await syncManifest(this.serverUrl, manifest.did, manifest, this.token);
  }

  async load(did: string): Promise<ClientSoulManifest | null> {
    return getLatestManifest(this.serverUrl, did, this.token);
  }

  async delete(_did: string): Promise<void> {
    // No server delete endpoint yet — no-op.
  }

  async listDids(): Promise<string[]> {
    // The server soul API is queried by DID; enumeration isn't exposed.
    return [];
  }
}
